src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -906,6 +900,145 @@ static bool usesReceiver(BoundExpression argument)
+
+        // This may introduce some temps into an outer scope. The caller has to deal with that.
+        [return: NotNullIfNotNull(nameof(receiver))]
+        private BoundExpression? AdjustReceiverForExtensionsIfNeeded(BoundExpression? receiver, Symbol member,
+            ArrayBuilder<BoundExpression>? storesOpt, [NotNullIfNotNull(nameof(tempsOpt))] ref ArrayBuilder<LocalSymbol>? tempsOpt)
+        {
+            if (receiver is null)
+            {
+                return receiver;
+            }
+
+            Debug.Assert(member is not null);
+            if (receiver is BoundTypeExpression)
+            {
+                return receiver;
+            }
+
+            if (receiver is BoundThisReference)
+            {
+                // PROTOTYPE this was only minimally tested to observe side-effects but needs more work (in CheckValue for instance, interpolation handlers)
+                // In an extension type, `this.UnderlyingTypeMember` is replaced with `this.<UnderlyingInstanceField>$.UnderlyingTypeMember`
+                Debug.Assert(receiver.Type is not null);
+                if (receiver.Type.IsExtension && !member.ContainingType.IsExtension)
+                {
+                    var thisExtensionType = (SourceExtensionTypeSymbol)receiver.Type;
+                    Debug.Assert(thisExtensionType.UnderlyingInstanceField is not null);
+                    if (!thisExtensionType.UnderlyingInstanceField.Type.Equals(member.ContainingType, TypeCompareKind.AllIgnoreOptions))
+                    {
+                        throw ExceptionUtilities.UnexpectedValue((thisExtensionType, member.ContainingType));
+                    }
+
+                    var oldSyntax2 = _factory.Syntax;
+                    _factory.Syntax = receiver.Syntax;
+
+                    var field = _factory.Field(receiver, thisExtensionType.UnderlyingInstanceField);
+
+                    _factory.Syntax = oldSyntax2;
+                    return field;
+                }
+            }
+
+            if (member.ContainingType is not { IsExtension: true } memberExtensionType)
+            {
+                return receiver;
+            }
+
+            Debug.Assert(receiver.Type is not null);
+            if (receiver.Type.IsExtension)
+            {
+                return receiver;
+            }
+
+            var oldSyntax = _factory.Syntax;
+            _factory.Syntax = receiver.Syntax;
+
+            tempsOpt ??= ArrayBuilder<LocalSymbol>.GetInstance();
+
+            // PROTOTYPE we'll want to allow extension member lookup and receiver adjustment for type parameter values
+            Debug.Assert(!receiver.Type.IsTypeParameter());
+            var refKind = GetRefKindForCapturedReceiverTemp(receiver, useNoneForReferenceType: true);
+            // PROTOTYPE if we decide to allow readonly members in extensions, we'll need to revisit this
+            if (refKind is RefKind.RefReadOnly)
+            {
+                refKind = RefKind.Ref;
```

Comment:
:memo: At the moment the emit layer (`EmitAddress`) fails when passing a temp with refKind `RefReadOnly` to `Unsafe.As(ref)` because that would create a copy. Instead, we force the copy here (during the assignment).
If/when we decide to support `readonly` extension members, we'll have to do something to re-interpret a `ref readonly UnderlyingType` into a `ref readonly Extension`.